### PR TITLE
Allow `Player` to proceed to results screen after failing

### PR DESCRIPTION
**Pull request by hlysine**
Split from #12877 , but rewritten to fit the new `Player`.

This PR adds a bool property `DisplayResultsOnFail` in `IApplicableFailOverride`, which instructs the `Player` to proceed to the results screen after failing, instead of showing the fail overlay.

When proceeding to the results screen after failing:
- The fail animation is played without frequency adjustments.
- The results screen is displayed after a short delay upon failing. It does not wait for the fail animation to complete.
- The music continues to play on the results screen.

### CODE REVIEW

File: osu.Game/Screens/Play/FailAnimation.cs [hunk 1]
```diff
@@ -28,7 +28,7 @@ public class FailAnimation : Component
 
         private readonly DrawableRuleset drawableRuleset;
 
-        private readonly BindableDouble trackFreq = new BindableDouble(1);
```
Comment (peppy) [inline osu.Game/Screens/Play/FailAnimation.cs:31]: This should not be changed (avoid null values wherever possible). Instead just don't apply the transform.

File: osu.Game/Screens/Play/FailAnimation.cs [hunk 2]
```diff
@@ -54,21 +54,26 @@ private void load(AudioManager audio, IBindable<WorkingBeatmap> beatmap)
         /// Start the fail animation playing.
         /// </summary>
         /// <exception cref="InvalidOperationException">Thrown if started more than once.</exception>
-        public void Start()
+        public void Start(bool adjustTrackFrequency)
```
Comment (peppy) [inline osu.Game/Screens/Play/FailAnimation.cs:57]: New parameter is not documented.

File: osu.Game/Screens/Play/Player.cs [hunk 3]
```diff
@@ -655,7 +657,7 @@ private void scoreCompletionChanged(ValueChangedEvent<bool> completed)
             // Currently, even if this scenario is hit, prepareScoreForDisplay has already been queued (and potentially run).
             // In scenarios where rewinding is possible (replay, spectating) this is a non-issue as no submission/import work is done,
             // but it still doesn't feel right that this exists here.
-            if (!completed.NewValue)
+            if (!completed.NewValue && !HealthProcessor.HasFailed)
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs:660]: Why?

File: osu.Game/Screens/Play/Player.cs [hunk 4]
```diff
@@ -133,6 +133,8 @@ public abstract class Player : ScreenWithBeatmapBackground, ISamplePlaybackDisab
         /// </summary>
         protected virtual bool CheckModsAllowFailure() => Mods.Value.OfType<IApplicableFailOverride>().All(m => m.PerformFail());
 
+        private bool displayResultsOnFail() => Mods.Value.OfType<IApplicableFailOverride>().Any(m => m.DisplayResultsOnFail);
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs:136]: Rename to `CheckModsDisplayResultsOnFail`

File: osu.Game/Screens/Play/Player.cs [hunk 5]
```diff
@@ -781,17 +786,25 @@ private bool onFail()
...
+            failAnimation.Start(!displayResultsOnFail());
 
             if (Mods.Value.OfType<IApplicableFailOverride>().Any(m => m.RestartOnFail))
                 Restart();
 
+            if (displayResultsOnFail())
+            {
+                scoreCompletionChanged(new ValueChangedEvent<bool>(false, true));
+            }
+
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs:798]: I don't even know where to start on this.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
This needs tests.

**Comment by smoogipoo** [comment]
I've been thinking for a while since the original PR, and I don't know if I really agree with this conceptually. I always found it odd and annoying that target practice finishes gameplay as soon as you miss an object. Eventually I believe we want to have a way to view the results screen for normal fails too so that you can see the hit statistics and share screenshots of your score probably with the gameplay progress bar/graph at the bottom.

**Comment by peppy** [comment]
> I've been thinking for a while since the original PR, and I don't know if I really agree with this conceptually. I always found it odd and annoying that target practice finishes gameplay as soon as you miss an object. Eventually I believe we want to have a way to view the results screen for normal fails too so that you can see the hit statistics and share screenshots of your score with the gameplay progress bar/graph at the bottom.

Completely agree with this for the general case. Should probably be a button on the fail screen to "show incomplete results" or similar.

That said, target practice was intended to be an endurance tester, where you play until you can't keep up (with the intention being that you generally don't get to the end of the beatmap ever). Not too sure how this should be handled. If we had the ability to dynamically update hitobject content we could remove all hitobjects after the first miss, or some other strategy, but not sure yet.

I think implementing with it just failing would be a good starting point, and don't necessarily see this PR being a prerequisite. Especially since in its current state it's adding more complexity to an area I'm actively working to reduce complexity and expand functionality.

**Comment by Henry-YSLin** [comment]
I also find it odd that gameplay just abruptly stops when I have 1 miss in target practice, but since the mod is designed to fail the player most of the time, it would be pointless if the player gets no feedback at all and only gets a fail overlay. So I created this PR to at least give the player some stats after gameplay. I never intend for changes in this PR to live long anyway, since there are a lot of broken stuff (pp, replay, etc) and target mod is the only mod using this functionality.

**Comment by Henry-YSLin** [comment]
That being said, if you think it is fine to just show a fail overlay for now, I can simply close this PR and revert related changes in #12877

**Comment by smoogipoo** [comment]
Let's show the fail overlay for now and keep the logic simple.

**Pull request closed without merging**